### PR TITLE
feat: Resource base class

### PR DESCRIPTION
**Pull request by eladb**
Added `cdk.Resource` and `cdk.IResource` (currently placeholders)
as bases for all constructs that represent AWS resources.

awslint rules:
- resource-class-extends-resource
- resource-interface-extends-resource

Apply across the library.

BREAKING CHANGE: `cloudformation.CustomResource` no longer extends CfnCustomResource.
* `ssm.ParameterProps` renamed to `ssm.ParameterOptions`.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/aws-apigateway/lib/deployment.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import cdk = require('@aws-cdk/cdk');
+import { Construct, DeletionPolicy, Resource, Token } from '@aws-cdk/cdk';
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-apigateway/lib/deployment.ts]: I personally don't like doing that, and it introduces a large amount of needless touches in those files...

File: packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts [hunk 2]
```diff
@@ -62,53 +62,34 @@ export interface CustomResourceProps {
...
-      throw new Error(`Custom resource type length > 60 (${resourceType})`);
-    }
-
-    if (!/^[a-z0-9_@-]+$/i.test(typeName)) {
-      throw new Error(`Custom resource type name can only include alphanumeric characters and _@- (${typeName})`);
-    }
-
-    this.addOverride('Type', resourceType);
+  public get ref() {
+    return this.resource.ref;
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts]: I don't think CustomResources even have a `Ref`.

File: packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts [hunk 3]
```diff
@@ -126,3 +107,24 @@ function uppercaseProperties(props: Properties): Properties {
...
+  }
+
+  const typeName = resourceType.substr(resourceType.indexOf('::') + 2);
+  if (typeName.length > 60) {
+    throw new Error(`Custom resource type length > 60 (${resourceType})`);
+  }
+
+  if (!/^[a-z0-9_@-]+$/i.test(typeName)) {
+    throw new Error(`Custom resource type name can only include alphanumeric characters and _@- (${typeName})`);
+  }
```
Comment (RomainMuller) [inline packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts]: Could have checked this whole thing in a single pass:

```
resourceType.match(/^Custom::[a-z0-9_@-]{1,60}$/);
```

File: packages/@aws-cdk/cdk/lib/resource.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+ */
+// tslint:disable-next-line:no-empty-interface
+export interface IResource extends IConstruct {
+
+}
+
+/**
+ * A construct which represents an AWS resource.
+ */
+export class Resource extends Construct implements IResource {
```
Comment (RomainMuller) [inline packages/@aws-cdk/cdk/lib/resource.ts]: `abstract`?

File: packages/@aws-cdk/aws-apigateway/lib/deployment.ts (hunk 1 above)
Comment (eladb) [reply packages/@aws-cdk/aws-apigateway/lib/deployment.ts]: OK.

File: packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts (hunk 3 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts]: This was moved, I prefer not to dig into it right now.

File: packages/@aws-cdk/cdk/lib/resource.ts (hunk 4 above)
Comment (eladb) [reply packages/@aws-cdk/cdk/lib/resource.ts]: Yap, good call.

File: packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts (hunk 2 above)
Comment (eladb) [reply packages/@aws-cdk/aws-cloudformation/lib/custom-resource.ts]: I'll remove

**Comment by jogold** [comment]
@eladb will this be updated https://github.com/awslabs/aws-cdk/blob/master/design/aws-guidelines.md?

Why are `ImportedXxx` classes not extending `Resource` (but sometimes well because they extend `XxxBase` which extends `Resource`)?

Thanks

**Comment by eladb** [comment]
Yes, we are finalizing the guidelines and will update the guidelines with all the new goodies, including linters for most of the guidelines.

**Comment by eladb** [comment]
In the meantime, here's the [draft](https://gist.github.com/eladb/6c74488f00eba60b02f83ae0c002e9f9)

**Pull request merged**